### PR TITLE
Vendoring libnetwork 31139cdb513aea5ad1ed08b60d4350a68b4c96db

### LINKED ISSUE #15081: [master]: ^C daemon causes panic
here are daemon logs: https://gist.github.com/anonymous/54f6dc3a00a8717fa3b9

havent tested RC but will do tomorrow

### PR DESCRIPTION
**Pull request by aboch**
Closes #15081 #14859 #14738 

(More test code present in libnetwork)

Signed-off-by: Alessandro Boch aboch@docker.com

### CODE REVIEW

**Comment by tiborvass** [comment]
@aboch sorry to be annoying, but would you mind putting the vendoring in a separate commit than the test?

**Comment by aboch** [comment]
@tiborvass Sure, I can do that. Just from my understanding, why is that needed ?

File: integration-cli/docker_cli_daemon_test.go [hunk 1]
```diff
@@ -780,6 +780,19 @@ func (s *DockerDaemonSuite) TestDaemonDefaultGatewayIPv4Explicit(c *check.C) {
 	deleteInterface(c, defaultNetworkBridge)
 }
 
+func (s *DockerDaemonSuite) TestDaemonDefaultGatewayIPv4ExplicitOutsideContainerSubnet(c *check.C) {
+	defaultNetworkBridge := "docker0"
+	deleteInterface(c, defaultNetworkBridge)
+
+	d := s.d
```
Comment (calavera) [inline integration-cli/docker_cli_daemon_test.go]: we don't really need to assign the daemon variable here :smile:

Comment (aboch) [inline integration-cli/docker_cli_daemon_test.go]: no we don't. Will remove.

**Comment by tiborvass** [comment]
@aboch it's to keep things clean. That way we understand that a vendoring commit cannot have changed the docker codebase and vice-versa. Just makes things nice and clean.
Also makes it very easy to review what changes in docker when the vendoring is big.

File: integration-cli/docker_cli_daemon_test.go [hunk 2]
```diff
@@ -780,6 +780,19 @@ func (s *DockerDaemonSuite) TestDaemonDefaultGatewayIPv4Explicit(c *check.C) {
...
+
+	// Program a custom default gateway outside of the container subnet, daemon should accept it and start
+	err := d.StartWithBusybox("--bip", "172.16.0.10/16", "--fixed-cidr", "172.16.1.0/24", "--default-gateway", "172.16.0.254")
+	c.Assert(err, check.IsNil)
+	defer deleteInterface(c, defaultNetworkBridge)
```
Comment (aboch) [inline integration-cli/docker_cli_daemon_test.go]: neither we need the `defer`. Will remove

**Comment by aboch** [comment]
@tiborvass Moved test code addition to a separate commit. 
@calavera Took care of your comment.

File: integration-cli/docker_cli_daemon_test.go [hunk 3]
```diff
@@ -780,6 +780,18 @@ func (s *DockerDaemonSuite) TestDaemonDefaultGatewayIPv4Explicit(c *check.C) {
 	deleteInterface(c, defaultNetworkBridge)
 }
 
+func (s *DockerDaemonSuite) TestDaemonDefaultGatewayIPv4ExplicitOutsideContainerSubnet(c *check.C) {
```
Comment (tiborvass) [inline integration-cli/docker_cli_daemon_test.go]: @aboch test should not pass on master

Comment (aboch) [inline integration-cli/docker_cli_daemon_test.go]: Issue #14859, besides the original one, tracks a [second issue](https://github.com/docker/docker/issues/14859#issuecomment-125124552).

Fix for the former issue is already present in Docker master, and this PR adds a test to cover that.
Fix for the second issue is part of this vendoring. Test code for it is in libnetwork.

So it is expected this integ test already passes on master.

Comment (tiborvass) [inline integration-cli/docker_cli_daemon_test.go]: Thanks for clarifying, where is the fix for the first issue?

Comment (aboch) [inline integration-cli/docker_cli_daemon_test.go]: You can find the link in my previous comment, I edited.

Comment (tiborvass) [inline integration-cli/docker_cli_daemon_test.go]: Confirmed the test fails if I checkout 5bdd4d0ec46ab071bbb633dde60b97b28911ade6^ (commit before #14966 's merge commit)

**Comment by tiborvass** [comment]
LGTM

**Comment by calavera** [comment]
LGTM

**Comment by calavera** [comment]
cherry picked into #15091

**Comment by mavenugo** [comment]
Fixes #14908

**Pull request merged**